src/contracts/core/Slasher.sol

```diff
@@ -1,102 +1,368 @@
+     */
+    function getWithdrawabilityAndScalingFactorAtEpoch(
+        address operator,
+        IStrategy strategy,
+        uint32 epoch
+    ) external view returns (bool, uint64) {
+        bool __canWithdraw = true;
+        uint64 scalingFactor = SlashingAccountingUtils.SHARE_CONVERSION_SCALE;
+        (uint32 lookupEpoch, bool found) = _getLookupEpoch(operator, strategy, epoch);
+        if (found) {
+            __canWithdraw = _canWithdraw(operator, strategy, lookupEpoch);
+            scalingFactor = slashingRequests[operator][strategy][lookupEpoch].scalingFactor;
+        }
+        return (__canWithdraw, scalingFactor);
+    }
 
-    function canSlash(address, address) external view returns (bool) {}
+    /**
+     * @notice gets whether withdrawals of the given strategy delegated to the given operator can be withdrawn
+     * @param operator the operator the withdrawal is delegated to
+     * @param strategy the strategy the withdrawal is from
+     * @param epoch the last epoch the withdrawal was slashable until
+     * @return whether the withdrawal can be executed
+     */
+    function canWithdraw(address operator, IStrategy strategy, uint32 epoch) public view returns (bool) {
+        bool __canWithdraw = true;
+        (uint32 lookupEpoch, bool found) = _getLookupEpoch(operator, strategy, epoch);
+        if (found) {
+            // if the slashing request is not executed, the operator cannot withdraw
+            __canWithdraw = _canWithdraw(operator, strategy, lookupEpoch);
+        }
+        return __canWithdraw;
+    }
 
-    function contractCanSlashOperatorUntilBlock(
-        address,
-        address
-    ) external view returns (uint32) {}
+    /**
+     * @notice gets the scaling factor for the given operator and strategy
+     * @param operator the operator to get the scaling factor for
+     * @param strategy the strategy to get the scaling factor for
+     * @return the scaling factor for the given operator and strategy
+     */
+    function shareScalingFactor(address operator, IStrategy strategy) public view returns (uint64) {
+        uint64 scalingFactor = _shareScalingFactor[operator][strategy];
+        if (scalingFactor == 0) {
+            return SlashingAccountingUtils.SHARE_CONVERSION_SCALE;
+        } else {
+            return scalingFactor;
+        }
+    }
 
-    function latestUpdateBlock(address, address) external view returns (uint32) {}
+    // includes pending but unexecuted slashings
+    function pendingShareScalingFactor(address operator, IStrategy strategy) public view returns (uint64) {
+        uint64 pendingScalingFactor = SlashingAccountingUtils.findNewScalingFactor({
+            scalingFactorBefore: _shareScalingFactor[operator][strategy],
+            // TODO: this particular lookup could potentially get more complex? e.g. also including the trailing epoch's pending amounts
+            rateToSlash: slashingRequests[operator][strategy][EpochUtils.currentEpoch()].slashingRate
```

Comment:
seems this should indeed be more complex -- we should presumably look up at least the previous epoch's value as well?